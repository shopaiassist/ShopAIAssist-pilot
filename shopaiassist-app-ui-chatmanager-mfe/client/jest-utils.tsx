import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
