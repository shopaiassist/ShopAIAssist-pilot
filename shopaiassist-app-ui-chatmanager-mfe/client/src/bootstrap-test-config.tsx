import { CSSProperties } from 'react';
import { MfeContext, UserAuth } from 'react';

import { MattersProps } from './App';

// In this test harness, mark dummy components we expect to be provided externally at runtime with an ugly recognizable
// border.
const ExternalComponent = ({ children, style }: { children: React.ReactNode; style?: CSSProperties }) => (
  <div
    style={{
      display: 'flex',
      width: '50%',
      border: `0.125rem solid purple`,
      gap: '1rem',
      padding: '1rem',
      ...style
    }}
  >
    {children}
  </div>
);

export const TEST_PROPS: MattersProps = {
  context: {
    user: {
      firstName: 'Hunter',
      lastName: 'Nenneau'
    },
    getAuthToken: () => {
      return new Promise((resolve) => {
        resolve({
          /* eslint-disable  @typescript-eslint/no-explicit-any */
          tokenType: 'op-orch' as any,
          productId: 'SHOPAIASSIST',
          bearerToken:
            ''
        } as UserAuth);
      });
    }
  } as MfeContext,
  createChatComponent: ({ chatIdentifier, chatFolderIdentifier, createPersistentChat }) => (
    <ExternalComponent style={{ width: '100%', height: '100%' }}>
      {chatIdentifier ? (
        `Chat Component for chat ${chatIdentifier}${chatFolderIdentifier ? ' within ' + chatFolderIdentifier : ''}.`
      ) : (
        <button onClick={() => createPersistentChat(chatFolderIdentifier)}>Button for new chat</button>
      )}
    </ExternalComponent>
  ),
  createFileManagementComponent: ({ chatIdentifier, chatFolderIdentifier, fileCollectionId, isMatterArchived }) => (
    <ExternalComponent>
      File Management Component for collection {fileCollectionId} (
      {chatIdentifier ? 'chat ' + chatIdentifier + ' ' : ''}
      {chatFolderIdentifier ? 'within folder ' + chatFolderIdentifier : ''}) that is{' '}
      {isMatterArchived ? 'archived' : 'not archived'}
    </ExternalComponent>
  ),
  isOpen: true,
  closeSidebar: () => {}
};
