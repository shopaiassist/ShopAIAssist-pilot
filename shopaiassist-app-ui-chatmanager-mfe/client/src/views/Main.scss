@import "../utils/styles/imports";

.main-app-container {
   display: flex;
   // This is the key to making the ai assistant respect the bounds of the parent container
   max-height: 100%;

   .main-body-container {
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;

      .main-header {
         display: flex;
         box-sizing: border-box;
         height: 3.688rem;
         padding: 0 1.5rem;
         align-items: center;
         justify-content: space-between;
         flex-shrink: 0;
         border-bottom: 1px solid var(--gray-gray-300-dedede, #DEDEDE);
         font-size: 1rem;

         .main-header-actions {
            align-items: center;
            display: flex;
            flex-direction: row;
         }
      }

      saf-breadcrumb-item {
         saf-icon {
            margin-right: 0.5rem;
         }

         align-self: center;
         &::part(control) {
            font-size: 1rem;
         }
         &:last-child::part(control) {
            pointer-events: none;
            cursor: default;
            text-decoration: none;
            color: black;
            font: $saf-type-body-default-md-strong-standard;
         }
      }

      .matters-ai-assistant-container {
         width: 100%;
         height: calc(100% - rem-calc(60));
      }
   }

   // // This selector is probably not really what we want.
   // & > div {
   //    height: 100%;
   // }

   .delete-dialog-footer {
      display: flex;
      justify-content: flex-end;
      gap: 1rem;
   }

   .matter-body-container {
      flex: 1;
   }

   .chat-title {
      margin-top: 0;
      margin-bottom: 0;
   }
}
