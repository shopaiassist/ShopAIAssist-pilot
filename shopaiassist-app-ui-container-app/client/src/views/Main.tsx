import React from 'react';
import { MfeContext } from '@';
import { useOutletContext } from 'react-router-dom';

import MfeContent from '../components/mfes/MfeContent';

const Main = () => {
  const context = useOutletContext<MfeContext | null>();

  return (
    <>
      {!!context && (
        <>
          <MfeContent context={context} />
        </>
      )}
    </>
  );
};

export default Main;
