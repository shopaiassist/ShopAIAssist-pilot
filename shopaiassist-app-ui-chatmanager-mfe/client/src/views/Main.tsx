import React, { ReactNode, useState } from 'react';
import { MfeContext } from 'react';
import MatterSidebar from '../components/matter/sidebar/MatterSidebar';
import useMainView from '../hooks/useMainView';
import { MattersProps } from '../App';
import { ChatActionsDialog, ChatOrFolderMenu, MatterDetailsDialog, TypewriterText } from '../components/common';
import MatterSettings from './MatterSettings/MatterSettings';
import { Breadcrumbs } from '@reacttext/prometheus';

import './Main.scss';
import { Text } from '@on/core-components/react';
import { TYPEWRITER_DELAY } from '../utils/constatnts';

interface MainProps extends Omit<MattersProps, 'context'> {
  context: MfeContext;
}

/**
 * Main component for the Matters application.
 *
 * This component manages the main view, including the sidebar, body, and dialogs for chat and matter details.
 *
 * @param {MainProps} props - The properties for the Main component.
 * @returns {JSX.Element} The rendered Main component.
 */
const Main = (props: MainProps): JSX.Element => {
  const { isOpen, closeSidebar, createChatComponent, createFileManagementComponent, createHeaderActions } = props;
  const [
    activeChat,
    activeFolder,
    chatsAndFolders,
    breadcrumbs,
    isMatterSettingsActive,
    isMatterFilesActive,
    createPersistentChat,
    setActiveChat
  ] = useMainView(props.context);
  // eslint-disable-next-line use-encapsulation/prefer-custom-hooks
  const [menuOpen, setMenuOpen] = useState(false);

  const returnMatterBodyComponent = () => {
    if (isMatterSettingsActive) {
      return <MatterSettings />;
    } else if (isMatterFilesActive) {
      if (activeFolder) {
        return createFileManagementComponent({
          chatIdentifier: activeChat?.treeItemId as string,
          chatFolderIdentifier: activeFolder.treeItemId,
          fileCollectionId: activeFolder.fileCollectionId,
          isMatterArchived: activeFolder.isArchived || false
        }) as ReactNode;
      }
    } else {
      return createChatComponent({
        chatFolderIdentifier: activeChat?.fileCollectionId,
        chatIdentifier: activeChat?.treeItemId,
        createPersistentChat
      }) as ReactNode;
    }
  };

  return (
    <div className="main-app-container">
      <MatterSidebar
        data-testid="matters-main-view-sidebar"
        tree={chatsAndFolders}
        onNewChat={() => setActiveChat(undefined)}
        isOpen={isOpen}
        closeSidebar={closeSidebar}
      />
      <div className="main-body-container">
        <div className="main-header">
          <h4 className='chat-title' id={`chat-text-${activeChat?.treeItemId}`}>
            <TypewriterText text={activeChat?.name || ''} delay={TYPEWRITER_DELAY} />
          </h4>
          {/* <Breadcrumbs data-testid="matters-main-view-breadcrumbs" items={breadcrumbs} /> */}
          {/* <div className="main-header-actions">
            {createHeaderActions?.() as React.ReactNode}
            {!!activeChat?.treeItemId && (
              <ChatOrFolderMenu
                isOpen={menuOpen}
                itemId={activeChat?.treeItemId}
                itemType="chat"
                toggleOpen={() => setMenuOpen(!menuOpen)}
              />
            )}
          </div> */}
        </div>
        <div className="matters-ai-assistant-container">{returnMatterBodyComponent()}</div>
      </div>

      <ChatActionsDialog />
      <MatterDetailsDialog />
    </div>
  );
};

export default Main;
