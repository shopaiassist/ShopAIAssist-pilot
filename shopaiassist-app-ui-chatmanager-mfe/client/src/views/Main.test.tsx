import { snapshot } from '../../jest-utils';
import Main from './Main';
import { TEST_PROPS } from '../bootstrap-test-config';
import React from 'react';

describe('Main View Tests', () => {
  const testProps = TEST_PROPS;

  describe('render', () => {
    it('should render', async () => {
      snapshot(<Main {...testProps} />);
    });
  });
});
