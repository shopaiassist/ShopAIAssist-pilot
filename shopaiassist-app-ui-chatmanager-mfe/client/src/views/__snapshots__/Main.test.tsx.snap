// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main View Tests render should render 1`] = `
<div
  class="main-app-container"
>
  <div
    class="matter-sidebar-container"
    style="transform: translateX(0px); margin-right: 0px;"
  >
    <div
      class="matter-sidebar"
      data-testid="matter-sidebar"
      is-open="true"
    >
      <div
        class="matter-sidebar-nav-header"
      >
        <span
          appearance="body-strong-md"
        >
          Home
        </span>
        <button
          appearance="tertiary"
          data-testid="matter-sidebar-nav-close"
        >
          <div
            iconname="close"
            size="16"
          />
        </button>
      </div>
      <div
        class="matter-body-header matter-sidebar-header"
        data-testid="matter-body-header"
      >
        <div
          class="new-folder-or-chat-button-container"
          data-testid="new-folder-or-chat-button-container"
        >
          <button
            appearance="primary"
            data-testid="new-chat-button"
          >
            <div
              class="new-folder-or-chat-button"
            >
              <div
                iconname="plus"
              />
              New
            </div>
          </button>
        </div>
        <div
          class="new-folder-or-chat-button-container"
          data-testid="sort-option-container"
        >
          <button
            appearance="tertiary"
            data-testid="sort-option-button"
          >
            <div
              class="new-folder-or-chat-button"
            >
              Sort by date
              <div
                iconname="angle-down"
              />
            </div>
          </button>
        </div>
      </div>
      <div
        class="chat-list-container"
      >
        <div
          class="empty-chat-list"
        >
           Chats will appear here
        </div>
      </div>
      <div
        class="matter-sidebar-footer"
      >
        <hr
          class="matter-sidebar-footer-separator"
        />
        <div
          class="matter-sidebar-archived-matters-selectable"
        >
          <div
            class="matter-sidebar-archived-matters-selectable-section"
          >
            <div
              class="selectable-item"
              data-testid="archived-matters-selectable"
            >
              <div
                class="matter-sidebar-archived-matters-selectable-item"
              >
                <div>
                  <div
                    appearance="light"
                    iconname="box-archive"
                    size="16"
                  />
                  <span
                    appearance="body-strong-md"
                    id="chat-archived-matters-navigation"
                  >
                    Archived Matters
                  </span>
                </div>
                <div
                  appearance="light"
                  class="matter-sidebar-archived-matters-selectable-icon"
                  iconname="chevron-right"
                  size="16"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="main-body-container"
  >
    <div
      class="main-header"
    >
      <saf-breadcrumb
        role="navigation"
      >
        <saf-breadcrumb-item>
          My Work
        </saf-breadcrumb-item>
      </saf-breadcrumb>
      <div
        class="main-header-actions"
      />
    </div>
    <div
      class="matters-ai-assistant-container"
    >
      <div
        style="display: flex; width: 100%; border: 0.125rem solid purple; gap: 1rem; padding: 1rem; height: 100%;"
      >
        <button>
          Button for new chat
        </button>
      </div>
    </div>
  </div>
  <div
    class="chat-actions-dialog"
    dialogtitle=""
    hidden=""
    isfooter="true"
  >
    <div
      class="chat-actions-dialog-body"
    >
      <div
        class="chat-rename-field"
      >
        <span
          appearance="heading-md"
        >
          * indicates required field
        </span>
        <textarea
          data-testid="chat-name-field"
          label="Name*"
          maxlength="128"
        />
      </div>
    </div>
    <div
      class="delete-dialog-footer"
      slot="footer"
    >
      <button
        appearance="secondary"
        data-testid="chat-dialog-cancel"
      >
        Cancel
      </button>
      <button
        appearance="primary"
        data-testid="chat-dialog-action"
        disabled=""
      >
        No translation: 
      </button>
    </div>
  </div>
  <div
    class="matter-details-dialog"
    dialogtitle="Create matter folder"
    hidden=""
    isfooter="true"
  >
    <div
      class="matter-details-dialog-body"
    >
      <span
        appearance="heading-md"
      >
        * indicates required field
      </span>
      <textarea
        data-testid="matter-dialog-name-field"
        label="Name*"
        maxlength="128"
      />
      <textarea
        data-testid="matter-dialog-description-field"
        label="Description"
        maxlength="2000"
        resize="vertical"
      />
      <textarea
        data-testid="matter-dialog-id-field"
        label="Matter ID"
        maxlength="128"
      />
    </div>
    <div
      class="matter-details-dialog-footer"
      slot="footer"
    >
      <button
        appearance="secondary"
        data-testid="matter-dialog-close"
      >
        Cancel
      </button>
      <button
        appearance="primary"
        data-testid="matter-dialog-confirm"
        disabled=""
      >
        Create matter folder
      </button>
    </div>
  </div>
</div>
`;
